### PR TITLE
hide levels of all types in locked stages

### PR DESCRIPTION
**Pull request by davidsbailey**
### Background

Fixes [LP-230](https://codedotorg.atlassian.net/browse/LP-230) Ensure locked Markdown levels can't be visited by direct url

### Description

Moves the logic for whether to display `#locked-stage` in place of level contents from partial `_level_group` out to `levels/show`.

### Testing

Adds new rails integrations tests and augments existing UI tests. What's interesting is that a lot of the integration tests cover things that I would normally have written as UI tests, but are better as integration tests because they are faster, cheaper and more reliable that way.

Other interesting note: the new rails tests are integration tests rather than controller tests because controller tests only let you specify request params (e.g. `{script_name: 'foo', stage_id: 3, ...}`, but the controller logic for accessing a page of a level group depends on the actual request path (e.g. `/s/foo/stage/3/...`) here: https://github.com/code-dot-org/code-dot-org/blob/c81319dc115e4e0dfac0cd9051a377acb884b9be/dashboard/app/controllers/script_levels_controller.rb#L124-L127 integration tests don't run into this problem because they let you specify the actual request path.

### CODE REVIEW

File: dashboard/test/integration/stage_lock_test.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    assert_select "#locked-stage", 1
+    # data-hidden indicates that the client will decide whether the teacher
+    # will see the locked-stage message via the ViewAsToggle.
+    assert_select "#locked-stage[data-hidden]", 1
+
+    # This needs to be an integration test rather than a controller test in
+    # order to follow the redirect which adds the /page/1 suffix.
+    get build_script_level_path(@lockable_level_group_sl)
+    assert_response :redirect
+    get URI.parse(@response.redirect_url).path
```
Comment (maddiedierker) [inline dashboard/test/integration/stage_lock_test.rb]: there is a `follow_redirect!` method available in integration tests that you should be able to use here. here's an example:

https://github.com/code-dot-org/code-dot-org/blob/94e720152b82437ac9bfdf85c9a1d1a082130b04/dashboard/test/controllers/omniauth_callbacks_controller/google_oauth2_test.rb#L53-L57

**Review by maddiedierker — APPROVED** [review]
LGTM! great testing as well 🎉

File: dashboard/test/integration/stage_lock_test.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/test/integration/stage_lock_test.rb]: Good call Maddie! That's much better. 

Incidentally, in looking for this method, I discovered that we have a many instances of `ActionDispatch::IntegrationTest` in the `dashboard/test/controllers` directory. I'm not sure how conventional our `dashboard/test` directory layout is, but what we currently have seems to suggests that any new instances of `ActionDispatch::IntegrationTest` should go in `dashboard/test/integration`, not `dashboard/test/controllers`. Just a thing to keep an eye out for in code reviews I guess.

**Review by islemaster — APPROVED** [review]
Cool - it makes a lot of sense to just pull that logic up another layer.

**Comment by davidsbailey** [comment]
confirmed locally the test failure is fixed by eliminating the name collision on StageLockTest in https://github.com/code-dot-org/code-dot-org/pull/28385 . Merging now since this needs to ship today and merging now is the shortest path toward that.

**Pull request merged**